src/refract/JsonSchema.cc

```diff
@@ -108,93 +99,152 @@ namespace
-    so::Object& render_schema(so::Object& schema, const NumberElement& e, flags options);
-    so::Object& render_schema(so::Object& schema, const BooleanElement& e, flags options);
-    so::Object& render_schema(so::Object& schema, const MemberElement& e, flags options);
-    so::Object& render_schema(so::Object& schema, const ExtendElement& e, flags options);
-    so::Object& render_schema(so::Object& schema, const OptionElement& e, flags options);
-    so::Object& render_schema(so::Object& schema, const SelectElement& e, flags options);
-    so::Object& render_schema(so::Object& schema, const RefElement& e, flags options);
-    so::Object& render_schema(so::Object& schema, const HolderElement& e, flags options);
-    so::Object& render_schema(so::Object& schema, const IElement& e, flags options = 0);
-
-    so::Object make_schema(const IElement& e, flags options = 0);
-
-    so::Object& render_schema(so::Object& schema, const ObjectElement& e, flags options)
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const ObjectElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const ArrayElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const EnumElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const NullElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const StringElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const NumberElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const BooleanElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const MemberElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const ExtendElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const OptionElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const SelectElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const RefElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const HolderElement& e, flags options);
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const IElement& e, flags options = 0);
+
+    so::Object make_schema(const IElement& e, so::Object& defs, flags options = 0);
+
+    so::Object& render_schema(so::Object& schema, so::Object& defs, const ObjectElement& e, flags options)
     {
         assert(false); // TODO @tjanc@ not implemented
         return schema;
     }
 
-    so::Object& render_schema(so::Object& schema, const ArrayElement& e, flags options)
+    so::Object& render_schema(so::Object& s, so::Object& defs, const ArrayElement& e, flags options)
```

Comment:
Why change from a nice readable name like schema to s?

Reply:
Because with this commit the implementation needs to choose whether to write into the schema directly or whether to write into `definitions`. See calls to `choose_target`. I could rename `s` to `schema`, but then all occurences of `schema` below need to be replaced with something... e.g. `target`?